### PR TITLE
merge alt-l1 branch to develop

### PR DESCRIPTION
**Pull request by boyuan-chen**
## Overview

Merge alt-l1 branch to develop branch

## Changes

- Add a flag in l2geth to enable the node for alt l2s.
- Add subgraph supports in SDK and DTL for alt l1s
- Add core contracts for Alt L1s to the `packages/contracts/`
- Modify core deployer to support alt l1s
- Add boba contracts for Alt L2s to the `packages/boba/contracts
- Modify boba deployer to support alt L2s
- Add integration tests for alt l1s to `integration-tests/test/alt-l2` folder
- Add MaxGasLimit in batch submitter for moonbeam
- Add integration tests for all networks
- Fix minor issues

## Testing

/

### CODE REVIEW

File: go/batch-submitter/config.go [hunk 1]
```diff
@@ -121,6 +121,9 @@ type Config struct {
 	// batch submitter can accept
 	MaxL1GasPrice uint64
 
+	// MaxGasLimit is the maximum gas limit that the batch submitter can accept
+	MaxGasLimit uint64
```
Comment (boyuan-chen) [inline go/batch-submitter/config.go:125]: A new parameter for `gasLimit` is added to batch submitter to solve the problem from `Moonbeam` (This is from `alt-l1` branch)

File: l2geth/params/config.go [hunk 2]
```diff
@@ -469,12 +470,28 @@ func (c *ChainConfig) IsFeeTokenUpdate(num *big.Int) bool {
...
+	}
+	if c.ChainID.Cmp(OpMainnetChainID) == 0 {
+		return true
+	}
+	if c.ChainID.Cmp(OpGoerliChainID) == 0 {
+		return true
+	}
+	return false
+}
+
```
Comment (boyuan-chen) [inline l2geth/params/config.go:494]: A new flag is added in l2geth to determine whether we want to launch it for Alt L2s or not.

File: packages/boba/contracts/src/index.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+export * from './hardhat-deploy-ethers'
```
Comment (boyuan-chen) [inline packages/boba/contracts/src/index.ts:1]: The new functions are exported from `boba deployer` to simplify the process of loading the contracts.

```
export const deployBobaContractCore = async (
  name: string,
  args: any[],
  signer: any
) {}
export const getBobaContractAt = async (
  name: string,
  address: string,
  signer: any
) {}
```

File: packages/boba/contracts/deploy/000-Messenger.deploy.ts [hunk 4]
```diff
@@ -1,102 +1,40 @@
...
 
-  Factory__L1_Messenger = new ContractFactory(
-    L1_MessengerJson.abi,
-    L1_MessengerJson.bytecode,
+  L1CrossDomainMessengerFast = await deployBobaContract(
+    hre,
+    'L1CrossDomainMessengerFast',
+    [],
     (hre as any).deployConfig.deployer_l1
   )
```
Comment (boyuan-chen) [inline packages/boba/contracts/deploy/000-Messenger.deploy.ts:26]: To deploy boba contracts, you just need to add the contract name.

File: packages/boba/contracts/contracts/ERC721Bridges/L2NFTBridgeAltL1.sol [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+    )
+        internal
+        onlyWithBillingContract()
+    {
+        // Collect the exit fee
+        L2BillingContract billingContract = L2BillingContract(billingContractAddress);
+        require(msg.value >= billingContract.exitFee(), "Insufficient Boba amount");
+        (bool sent,) = billingContractAddress.call{gas: 3000, value: billingContract.exitFee()}("");
+        require(sent, "Failed to collect exit fee");
+
```
Comment (boyuan-chen) [inline packages/boba/contracts/contracts/ERC721Bridges/L2NFTBridgeAltL1.sol:378]: The billing for `L2NFTBridge` is fixed. It was disabled in the `alt-l1` branch.

File: integration-tests/test/eth-l2/boba_turing_credit.spec.ts [hunk 6]
```diff
@@ -1,21 +1,16 @@
...
+import { ethers } from 'hardhat'
 
 import { getContractFactory } from '@eth-optimism/contracts'
-
-import L1ERC20Json from '@boba/contracts/artifacts/contracts/test-helpers/L1ERC20.sol/L1ERC20.json'
-import L2GovernanceERC20Json from '@boba/contracts/artifacts/contracts/standards/L2GovernanceERC20.sol/L2GovernanceERC20.json'
-import TuringHelperJson from '../artifacts/contracts/TuringHelper.sol/TuringHelper.json'
-
-import TuringTestJson from '../artifacts/contracts/TuringTest.sol/TuringTest.json'
+import { getBobaContractAt } from '@boba/contracts'
```
Comment (boyuan-chen) [inline integration-tests/test/eth-l2/boba_turing_credit.spec.ts:8]: We don't need to load contracts from `.json`. It is replaced by functions called
```
getBobaContractAt()
getBobaContractABI()
deployBobaContractCore()
```

File: l2geth/core/state_transition.go [hunk 7]
```diff
@@ -404,6 +429,28 @@ func (st *StateTransition) refundGas() {
 	if st.isBobaFeeTokenSelect {
 		remainingETH := new(big.Int).Mul(new(big.Int).SetUint64(st.gas), st.msg.GasPrice())
 		remainingBoba := new(big.Int).Mul(remainingETH, st.bobaPriceRatio)
+		if !st.isEthereumL2 {
```
Comment (InoMurko) [inline l2geth/core/state_transition.go:432]: is this chunk tested anywhere?

Comment (boyuan-chen) [reply l2geth/core/state_transition.go:432]: Yes! https://github.com/bobanetwork/boba/blob/c43465e34416941cdef5d2f90e1403bf7fdf57bf/l2geth/boba/state_transition_test.go#L200

File: .circleci/config.yml [hunk 8]
```diff
@@ -234,10 +235,394 @@ jobs:
...
+      - run:
+          name: Generate cache key from all hardhat.config.ts files
+          working_directory: ops
+          command: ./scripts/checksum.sh /tmp/checksum.txt
+
+      - restore_cache:
+          keys:
+            - solc-compilers-{{ checksum "/tmp/checksum.txt" }}-{{ .Environment.CACHE_VERSION }}
+
+      - run:
```
Comment (InoMurko) [inline .circleci/config.yml]: this is a perfect usecase for https://circleci.com/docs/reusing-config/

Comment (boyuan-chen) [reply .circleci/config.yml]: Done!
```
  integration-tests:
    executor: intergration-tests-executor
    parallelism: 4
    steps:
      - download-solidity-compilers
      - run-integration-tests:
          docker_compose_file: "docker-compose.yml"
          test_folder: "eth-l2"
      - report-integration-tests
```

**Review by bobabot-enya-ai — APPROVED** [review]
as far as I can review the massive PR it looks OK.
I would much prefer the merge happen in steps, like service per service ... add gas price orcle changes, then l2geth, then teleportation etc etc.

I would really like @mmontour1306 to look at the changes in l2geth folder tho!

**Comment by InoMurko** [comment]
as far as I can review the massive PR it looks OK.
I would much prefer the merge happen in steps, like service per service ... add gas price orcle changes, then l2geth, then teleportation etc etc.

I would really like @mmontour1306 to look at the changes in l2geth folder tho!

**Review by InoMurko — APPROVED** [review]

File: l2geth/core/state_transition.go (hunk 7 above)
Comment (InoMurko) [reply l2geth/core/state_transition.go:432]: awesome, thank you

**Comment by InoMurko** [comment]
and @souradeep-das if you focus on the contracts added pls!

**Review by mmontour1306 — APPROVED** [review]

File: integration-tests/test/alt-l2/native-eth-ovm-calls.spec.ts [hunk 9]
```diff
@@ -0,0 +1,375 @@
+import { BigNumber, Contract, ContractFactory, Wallet } from 'ethers'
```
Comment (souradeep-das) [inline integration-tests/test/alt-l2/native-eth-ovm-calls.spec.ts]: minor: do we want to rename this file?

File: integration-tests/test/alt-l2/native-eth.spec.ts [hunk 10]
```diff
@@ -0,0 +1,344 @@
+/* Imports: External */
```
Comment (souradeep-das) [inline integration-tests/test/alt-l2/native-eth.spec.ts]: minor: do we want to rename this file?

File: packages/boba/contracts/contracts/LP/L2LiquidityPoolAltL1.sol [hunk 11]
```diff
@@ -0,0 +1,943 @@
...
+import "@eth-optimism/contracts/contracts/libraries/bridge/CrossDomainEnabled.sol";
+import "@eth-optimism/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol";
+
+/* External Imports */
+import '@openzeppelin/contracts/utils/math/SafeMath.sol';
+import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import "@eth-optimism/contracts/contracts/L2/predeploys/OVM_GasPriceOracle.sol";
+import "@eth-optimism/contracts/contracts/L2/messaging/L2StandardBridge.sol";
```
Comment (souradeep-das) [inline packages/boba/contracts/contracts/LP/L2LiquidityPoolAltL1.sol]: i guess it doesn't make a difference but lets use
L2StandardBridgeAltL1?

File: packages/boba/contracts/deploy/002-LiquidityPools.deploy.ts [hunk 12]
```diff
@@ -55,16 +40,14 @@ const deployFn: DeployFunction = async (hre) => {
   console.log(`L2LiquidityPool deployed to: ${L2LiquidityPool.address}`)
 
   console.log(`Deploying L1LP...`)
-  L1LiquidityPool = await Factory__L1LiquidityPool.deploy()
-
-  await L1LiquidityPool.deployTransaction.wait()
+  L1LiquidityPool = await deployBobaContract(
+    hre,
+    'L1LiquidityPool',
```
Comment (souradeep-das) [inline packages/boba/contracts/deploy/002-LiquidityPools.deploy.ts]: L1LiquidityPoolAltL1?

File: packages/boba/contracts/src/index.ts (hunk 3 above)
Comment (souradeep-das) [reply packages/boba/contracts/src/index.ts:1]: these are great!

File: packages/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol [hunk 13]
```diff
@@ -29,4 +29,12 @@ library Lib_PredeployAddresses {
     address internal constant PROXY__BOBA_GAS_PRICE_ORACLE =
         0x4200000000000000000000000000000000000024;
     address internal constant BOBA_GAS_PRICE_ORACLE = 0x4200000000000000000000000000000000000025;
+
```
Comment (souradeep-das) [inline packages/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol:32]: where are these located? where do we put them?
L2StandardBridgeAltL1
BobaTuringCreditAltL1
Boba_GasPriceOracleAltL1

**Review by souradeep-das — COMMENTED** [review]
great stuff!

File: packages/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol (hunk 13 above)
Comment (boyuan-chen) [reply packages/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol:32]: These contract addresses are the same as the Ethereum L2. In the process of generating the state dump file, those contract names are mapping to the correct one.

https://github.com/bobanetwork/boba/blob/c130e2b3d161d3a78bc0ff2dd1748eaf2f1b511e/packages/contracts/src/make-genesis.ts#L195-L232

https://github.com/bobanetwork/boba/blob/c130e2b3d161d3a78bc0ff2dd1748eaf2f1b511e/packages/contracts/src/make-genesis.ts#L307-L311

File: integration-tests/test/alt-l2/native-eth-ovm-calls.spec.ts (hunk 9 above)
Comment (boyuan-chen) [reply integration-tests/test/alt-l2/native-eth-ovm-calls.spec.ts]: Rename to native-boba-ovm-calls.spec.ts!

File: packages/boba/contracts/contracts/LP/L2LiquidityPoolAltL1.sol (hunk 11 above)
Comment (boyuan-chen) [reply packages/boba/contracts/contracts/LP/L2LiquidityPoolAltL1.sol]: Yes!

File: integration-tests/test/alt-l2/native-eth.spec.ts (hunk 10 above)
Comment (boyuan-chen) [reply integration-tests/test/alt-l2/native-eth.spec.ts]: Rename to native-boba.spec.ts!

File: packages/boba/contracts/deploy/002-LiquidityPools.deploy.ts (hunk 12 above)
Comment (boyuan-chen) [reply packages/boba/contracts/deploy/002-LiquidityPools.deploy.ts]: Good catch!

**Comment by boyuan-chen** [comment]
I did sync tests. The new DTL and l2geth successfully synced Boba Ethereum L2, Bobabase, BobaBnb and BobaAvax as replicas.

File: packages/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol (hunk 13 above)
Comment (souradeep-das) [reply packages/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol:32]: aah i see, great!

Comment (InoMurko) [reply packages/contracts/contracts/libraries/constants/Lib_PredeployAddresses.sol:32]: maze!

**Comment by InoMurko** [comment]
can we merge this?

**Review by souradeep-das — APPROVED** [review]

**Comment by boyuan-chen** [comment]
> can we merge this?

Yes!

**Pull request merged**